### ISSUE TITLE
spec: remove string(int)

### ISSUE THREAD

**Issue by robpike** [golang/go#3939]
<pre>The conversion of an integer to a string was put in very early to bootstrap formatted
printing, if I remember correctly. It's odd, though, and no longer necessary, if it ever
was. It also causes inconsistencies, since string(0xD800) cannot produce the UTF-8
encoding for that code point (by definition, surrogates are not legal in UTF-8) so must
produce something else. We chose the &quot;\uFFFD&quot; since that's the only reasonable
option, but that means:

1) the result isn't obvious
2) string(0xD800) and &quot;\uD800&quot; do different things: the former produces the
UTF-8 for U+FFFD while the latter is statically rejected.

I propose that, in some remote future, we eliminate this conversion from the language.</pre>

**Comment by dolmen** [comment]
I've been beaten by this "feature" in production code today.
My code was equivalent to this (but of course not so obvious: type, const and cast were in 3 different packages, and the cast result was an argument to a function, so that made the bad use of the cast quite hidden):
```go
   type XType int16
   const x XType = 0x0001
   var y = string(x)
```
I think that `go vet` should raise an alert at least for my case (conversion to string of a value of a named integer type).

**Comment by robpike** [comment]
It is valid code, so vet will not complain about it. Perhaps golint should, but not vet.

**Comment by urandom** [comment]
Would it be possible to make string(int), return the string representation of the int. More users will probably expect this behaviour:
```
a := 42
fmt.Print(string(a)) // "42"
```

**Comment by dominikh** [comment]
@urandom I would not expect an innocuously looking conversion to invoke an expensive number formatting routine. The solution to fixing a (supposedly) confusing construct isn't to replace it with another, equally confusing one.

**Comment by rsc** [comment]
No, that is not possible. It will silently break existing programs in a subtle way, for no real reason. If you want to convert an int (or a float or a bool) to a string, use package strconv. In particular, using strconv allows you more control over the exact formatting.

**Comment by griesemer** [comment]
Another argument in favor of abandoning this feature is that it behaves differently when shifts are involved. See #26096 .

**Comment by KindlyFire** [comment]
I wonder how to achieve what `string(int)` does with `strconv` ?

Doing `string(65)` returns `"A"`, but I can't find a function in `strconv` that does that.

**Comment by cznic** [comment]
`fmt.Sprintf("%c", 65)`

https://play.golang.org/p/xFVd63hT4JA

**Comment by griesemer** [comment]
`string(int)` is really quite a complex operation as it needs to know utf8 encoding. @robpike 's original comment in this issue explains the rationale for removing it. Now that we have the proper libraries (uniccode, utf8) we can write this code explicitly. See e.g. `runeToString` here: https://play.golang.org/p/ZnUF0Oc_dAG .

**Comment by magical** [comment]
@griesemer  You don't even have to use a library. This works: `string([]rune{65})` https://play.golang.org/p/I6aX3c6fG71. Unless we're proposing removing `string([]rune)` as well.

**Comment by griesemer** [comment]
@magical Good point! I don't know about removing string([]rune). String(int/rune) is particularly annoying because the argument may be shift operation which opens a can of worms (see #26096 and various related issues).

**Comment by mimoo** [comment]
Note that python allows you to do str(42) -> "42" so this might bring confusion also.

**Comment by ianlancetaylor** [comment]
I noticed that some existing code uses `string(os.PathSeparator)`, which would become invalid under this proposal.  Not a reason not to do it, just a note about consequences.

**Comment by networkimprov** [comment]
This would break code that uses character constants as switches/flags/arguments, which
a) can be included in a string (e.g. filename, URL), and
b) are meaningful when printed in a log.

**Comment by griesemer** [comment]
@networkimprov I'm not sure I follow your comment. Can you be a bit more specific (example)?

One place where `string(x)` cannot simply be replaced with a corresponding library call is where `x` is a constant and we rely on the fact that `string(x)` also is a constant:

```Go
const s = string(c)
```

This won't be possible anymore. In many cases, `const` can be replaced with `var`; in others it may be necessary to do the `string(c)` manually rather than have it done by the compiler.

**Comment by networkimprov** [comment]
I use string(char) like this:
```
const ( eThis = 'I'; eThat = 'A' ) // alternative to iota
...
aFileName := aDir +"/"+ string(eThis) + aName
...
func F(iFlag byte, iData T) {
   log.Println("F with ", string(iFlag))

   switch iFlag {
   case eThis: ...
   case eThat: ...
   }
}
```

**Comment by rsc** [comment]
In all the discussion of this proposal I'd been focusing in my head on

    i := 3
    s := string(i)

which is clearly wrong and not going to hurt to remove. But I hadn't thought much about

    c := 'x'
    s := string(c)

which is clearly fine and does happen. Worse, it happens in constants like `"abc"+string(filepath.Separator)`. 

#30614 proposes adding filepath.SeparatorString, but having to update all that code seems unfortunate.

We could make this issue about rejecting string(int) but not string(rune), but today string(rune) and string(int32) are indistinguishable, so string(int32) would slip in too.

Or perhaps we should put this issue (#3939) off until we decide #29012 (separate rune and int32).

**Comment by go101** [comment]
Is it ok to add a builtin `char` function to return one-rune string?

**Comment by griesemer** [comment]
I suppose one could add a new built-in, but `char` seems an unfortunate name for a function returning a string.

**Comment by ianlancetaylor** [comment]
I don't see any meaningful difference between a `char` builtin and the current `string` conversion.

**Comment by go101** [comment]
It has the benefit of avoid mistaking `stirng(aInt)` as the string representation of `aInt`, but not more others.

How about allow concentrating constant strings and runes?

**Comment by fohristiwhirl** [comment]
If you'll accept a piece of anecdata, I had in my code a bit of string generation morally equivalent to this:

Or see [Go Playground](https://play.golang.org/p/K6QDaVeIM1X):

```golang
foo := "字符串"
bar := ""

for i := 0; i < len(foo); i++ {
	bar += string(foo[i])
}
```

I had been treating `string(n)` as a way to convert bytes to strings of length 1, not realising it would generate something bigger if the byte was > 127. My mistake of course, but a real example of this biting someone...

**Comment by qrpnxz** [comment]
It's not totally clear to me if this proposal also removes `string(rune)`. I hope not since it's very useful for debugging and constants as griesemer noted.

I don't think removing it totally is the solution. If `string(rune)` using `utf8.RuneError` replacement isn't obvious to someone, it makes more sense to me that they accept that is the case rather than remove it. It is a stone written fact that strings are valid utf8. It behaves just like `string([]rune)` and `string([]byte)`, which behave just like `utf8.EncodeRune`, except that for the latter you have to add an import and write way more code.

It's totally fair to make `string(int)` have to be `string(rune)` however.

**Comment by ianlancetaylor** [comment]
I think it's clear that we have to continue to support `string(r)` where `r` has type `rune`.  And it's likely that we have to continue to support `string(b)` where `b` has type `byte` or, equivalently, `uint8`.  So what we are talking about removing is the conversions of other integer types to `string`.

Treating `rune` and `byte`/`uint8` as a special case is not as bad as it may sound; we already do exactly that for converting values of type `[]rune` and `[]byte` (but not slices of other integer types) to `string`.

**Comment by alanfo** [comment]
I am beginning to wonder whether removing the ```string(int)``` conversion is going to be more trouble than it's worth, even for integer types other than ```rune``` and ```byte```.

I see from #32479 that the intermediate step of adding a check for this to ```go vet``` or other code analyzers presents difficulties if the suggested fix breaks existing code when the integer value being converted overflows ```rune```.

Although one wouldn't expect this to occur in practice, people will no doubt have used these conversions in weird, unexpected but (to them) reasonable ways knowing precisely what the results would be.

Incidentally, just to reinforce the point that the ```string(byte)``` conversion should continue to be allowed, my attention was recently drawn (by a link posted on reddit) to  [this book](https://interpreterbook.com)  about writing an interpreter in Go. In the free sample chapter, on page 7, I noticed this function:

```go
func newToken(tokenType token.TokenType, ch byte) token.Token {
    return token.Token{Type: tokenType, Literal: string(ch)}
}
```
The intended audience for this book is people who know enough Go to be able to make sense of the code and I don't think those people are going to be impressed if the code in the book fails to compile (or vet).

**Comment by ianlancetaylor** [comment]
It's clearly essential to find how much working code that is correctly using the conversion will break.  We don't know that yet.

**Comment by ianlancetaylor** [comment]
In 1.15 we have a vet warning for this case (#32479).  Putting this issue on hold until we have more experience with that.  Also waiting until we can assume that everyone is using Go modules with a language version.

**Comment by griesemer** [comment]
We have had a vet check in place for a long time now, so there's likely little code that still depends on this feature. At the same time, there's no significant benefit to removing the mechanism (besides simplifying the spec and compiler a little bit). The [spec](https://golang.org/ref/spec#Conversions) also suggest that the feature not be used.

From a pragmatic perspective, there is little to gain from actually removing the feature, and a (if very small) chance of breaking existing code.

Therefore, this is a **likely decline**. Leaving open for four weeks for final comments.
Leaving to @alandonovan to add code analyzer results if appropriate.

— @gri, for the language proposal review group

**Comment by aclements** [comment]
I assume at this point we would tie it to the language version, not just drop it entirely, so old code wouldn't break at all. Having this produce a real compile error has definite advantages over a vet check. Vet is not universal.

**Comment by ohir** [comment]
> so there's likely little code that still depends on this feature

Not much of a public code.  Though I bet in archived old one-offs (for me it were mostly a single file builders/generators) this cast might still linger.

**Comment by griesemer** [comment]
Fair point. Making this version-specific seems like a good alternative to just leaving it in.

**Comment by neild** [comment]
The proposal is:

When the language version is go1.28 or higher, string(x) where x is an integer type other than byte or rune is no longer valid Go.

The reason for permitting byte and rune is because the existing go vet check permits these, and existing code often contains conversions such as string(os.PathSeparator).

Is that correct?

**Comment by griesemer** [comment]
This proposal was moved to the general proposal review, removing LanguageChangeReview and FinalCommentPeriod label.

**Comment by aclements** [comment]
This proposal has been added to the [active column](https://go.dev/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— aclements for the proposal review group

**Comment by TheCoreMan** [comment]
I'm for this suggestion.

While discussing it on the [Cup o' Go E168](https://cupogo.dev/), I've [searched public code using sourcegraph](https://sourcegraph.com/search?q=context:global+lang:Go+%5Cs%2Bstring%5C%28%5Cd%2B%5C%29+count:11000&patternType=regexp&case=yes&sm=0). This search doesn't cover all cases - it just covers the cases where literal digits have been passed, so misses very obvious cases like:

```go
i := 0
nullish := string(i)
```

I don't think there's a search that can find all cases without being AST aware, and I'll leave running `stringintconv` on all Go source code in the world to someone with a bit more compute 😅

In any case, seems like there are real usages, but obviously they can all be upgraded. Some simple cases I've seen:

* string(0) / NUL
* other control chars (like using `9` for TAB)
* Printable ASCII chars (like `string(65)`)
* Literal non-ASCII Unicode code point

So perhaps a conversion guide/modernizer/recommendation to use `stringintconv` is in order when this starts spitting out errors?

Given this data + the fact that this is language-version-specific, I think it's really safe to remove `string(int)` and I'm all for it! 👍

**Comment by aclements** [comment]
Proposal review discussion:

- Generally, this change should follow exactly what the vet check does today.
- What exactly happens with `string` of a numeric constant, like `string(0)`? The vet check specifically calls out `string` applied to an untyped int. We'd get an equivalent result at a "spec level" if we let `x` follow the usual type defaulting rules and then `string` rejects an `int` argument, though having the nicer error in the compiler like the vet error is valuable.
- We're all in favor of doing this. @griesemer will write the exact spec text (which should be pretty straightforward).

**Comment by aclements** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://go.dev/s/proposal-status#likely-accept)**.
— aclements for the proposal review group


The proposal is to restrict integer-to-string conversions to *only* `byte` or `rune`, starting in language version 1.28.

This restriction has long been implemented as a vet check. This proposal simply lifts it to the language level.

The detailed spec wording can be found in https://go.dev/cl/830044. The core change is simple (but could use some word-smithing):

```diff
 A non-constant value <code>x</code> can be converted to type <code>T</code>
 in any of these cases:

 ...

 	<li>
-	<code>x</code> is an integer or a slice of bytes or runes
-	and <code>T</code> is a string type.
+	<code>x</code> is an integer of type <code>byte</code> or <code>rune</code>
+	or an untyped rune constant
+	or a slice of bytes or runes and <code>T</code> is a string type.
 	</li>

 ...

 <h4 id="Conversions_to_and_from_a_string_type">Conversions to and from a string type</h4>

 ...

-Finally, for historical reasons, an integer value may be converted to a string type.
+Finally, an integer value of type <code>byte</code> or <code>rune</code>
+or an untyped rune constant
+may be converted to a string type.
```

**Comment by mateusz834** [comment]
Isn't it weird that:

```go
string(int32(1)) // does not work
string(rune(1)) // works

string(uint8(1)) // does not work
string(byte(1)) // works
```

knowing that:

```go
type rune = int32
type byte = uint8
```

`rune` should be identical to `int32`, right?

It would make sense to me if rune and byte were not aliases. But since they are aliases, I can see this be confusing.

**Comment by mateusz834** [comment]
Also note that:

```go
string([]int32{1, 2, 3})
string([]rune{1, 2, 3})
```

both work

**Comment by willfaught** [comment]
@mateusz834 Do you think that user type aliases to rune/byte should also be permitted?

**Comment by griesemer** [comment]
@mateusz834 ~Thanks for pointing this out - this is just a bug in the CL -~ `int32` and `rune` as well as `uint8` and `byte` should behave the same ~- they are aliases. Will fix.~

Correction: I'm not sure why you say the [examples](https://github.com/golang/go/issues/3939#issuecomment-5702577650) don't work - they do work with the proposed changes (and they also work in 1.27).    Anyway, I have added respective examples to the pending spec change and the implementation change.

Also, to be clear, when we talk about the type `byte` or `rune` in the spec, we always also mean `uint8` and `int32`, respectively - aliases may have different names but denote the same type.

**Comment by mateusz834** [comment]
> https://github.com/golang/go/issues/3939#issuecomment-5702577650 don't work

I have not checked the implementation, but only looked at the spec and thought that this case should be clarified better, also this discussion here did not mention the fact anywhere and thought that it might be worth bringing up so we are aware of this before this gets finally accepted. I got few times confused that `rune` and `int32` are in fact the same type (e.g with reflect/generics), thus  mentioned that, since i am aware that it is easy to miss that.

> I have added respective examples to the pending spec change and the implementation change.

Thanks

> Also, to be clear, when we talk about the type byte or rune in the spec, we always also mean uint8 and int32, respectively - aliases may have different names but denote the same type.

Oh I see, my initial thought was that we distinguish `rune` and `int32` as two different things in this, since, say vet check could do that (i.e. require an explicit conversion to rune), but for the compiler it might not make sense.

**Comment by griesemer** [comment]
@mateusz834 Ok, thanks for the explanations. Note that the spec is clear that aliases denote identical types (see the examples in [Alias Declarations](https://golang.org/ref/spec#Alias_declarations)). Anyway, with the additional examples this should be clear now.

**Comment by jub0bs** [comment]
Out of curiosity: is this not the first breaking change (as in "code that would compile with Go 1.27 may no longer compiler with Go 1.28") ever introduced by a new version of Go (post 1.0)?

**Comment by griesemer** [comment]
@jub0bs In a strict sense that is partly true (it's not the first, because we changed the semantics of loop iteration variables before). But go vet has checked exactly this for many years now, and that carve-outs in this proposal allow for exactly the cases where string(integer) still matters (like `string(os.PathSeparator)`). This has all been discussed in this issue.

I also note that the proposed new restriction only applies to Go 1.28 and newer. Thus, in the unlikely case that old code doesn't have a Go version and makes use of, say `string(65)`, it will either need to be fixed (`string(rune(65))`) or have a Go module version. The latter is typically trivial; the former may be trivial, too.

**Comment by zigo101** [comment]
> would compile with Go 1.27 may no longer compiler with Go 1.28"

This is not bad actually. At least you will be aware of the change.

On the contrary, you can't be aware of [some behavior changes caused by Go 1.27 3-clause-for loop semantics](https://go101.org/blog/2024-03-01-for-loop-semantic-changes-in-go-1.22.html) in time. Go team never provided a tool to detect such changes.

**Comment by aclements** [comment]
No change in consensus, so **[accepted](https://go.dev/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— aclements for the proposal review group


The proposal is to restrict integer-to-string conversions to *only* `byte` or `rune`, starting in language version 1.28.

This restriction has long been implemented as a vet check. This proposal simply lifts it to the language level.

The detailed spec wording can be found in https://go.dev/cl/830044. The core change is simple (but could use some word-smithing):

```diff
 A non-constant value <code>x</code> can be converted to type <code>T</code>
 in any of these cases:

 ...

 	<li>
-	<code>x</code> is an integer or a slice of bytes or runes
-	and <code>T</code> is a string type.
+	<code>x</code> is an integer of type <code>byte</code> or <code>rune</code>
+	or an untyped rune constant
+	or a slice of bytes or runes and <code>T</code> is a string type.
 	</li>

 ...

 <h4 id="Conversions_to_and_from_a_string_type">Conversions to and from a string type</h4>

 ...

-Finally, for historical reasons, an integer value may be converted to a string type.
+Finally, an integer value of type <code>byte</code> or <code>rune</code>
+or an untyped rune constant
+may be converted to a string type.
```
